### ISSUE TITLE
mutagen request: ability to wait for cache sync to complete

### ISSUE THREAD

**Issue by allenpetersen** [docker/for-mac#4641]
I'm testing out Docker Desktop for Mac 2.3.1.0 with mutagen caching.

Due to the eventual consistency of mutagen I'm seeing issues with some scripts in our development workflow.

We have some scripts that run commands in a docker container which will create or modify files. After the command finishes the next commands in the script expect to see the new or modified files. With Mutagen the files have not be created or still have the old state once the `docker run` command completes.

Something like `docker cache sync` which would block until mutagen is done syncing could be a possible solution.

**Comment by allenpetersen** [comment]
Here is a test script that shows it takes about 9 seconds on my system to sync a 50MB file.
```
#!/bin/bash

rm /some/cached/directory/cache-test

echo "Creating 50MB random file"
docker run -it --rm -v /some/cached/directory:/data ubuntu dd if=/dev/urandom of=/data/cache-test bs=1M count=50 status=none
echo "File created, waiting for sync"

SECONDS=0
while [ ! -f /some/cached/directory/cache-test ]
do
  printf "."
  sleep 0.5
done

echo " found file"
echo "It took $SECONDS seconds to sync the file."
```

output:
```
Creating 50MB random file
File created, waiting for sync
................. found file
It took 9 seconds to sync the file.
```

I don't hate the 9 second delay, that is better than the 5+ minutes it takes to run the build command without Mutagen. I just need to know when the sync is done to I can continue the rest of the script.

**Comment by fdanielsen** [comment]
Can confirm that this is an issue which would be nice to be able to handle in a good way. I was mostly testing this new Mutagen based feature in Docker out of curiosity today, though I have used Mutagen directly before with good experiences.

We have a script that runs tests and coverage reports inside a Docker container. After the coverage command has run inside the container the local script immediately tries accessing the report from the local filesystem, which it then can't find. This is a file of only about 0.5MB. Adding some delay trying to read the report in our script it seems it can take <3-5 seconds for it to become visible on the host filesystem.

Can't remember this having been a problem when I used Mutagen directly on the same project, so maybe there's some configuration of the consistency in Mutagen that one could be allowed changing in the Docker setup too?

**Comment by stephen-turner** [comment]
Thanks @allenpetersen and @fdanielsen, this is really good feedback. We're busy making a list of changes we want to make before this feature reaches Stable, so this discussion is timely.

Assign to @djs55.

**Comment by andig** [comment]
My personal feeling is that an invocation of `docker run` should end with consistent state. Requiring an explicit `cache sync` would make any script depend on the docker-internal implementation and likely break functionality for a larger portion of the user base?

**Comment by fdanielsen** [comment]
@stephen-turner No problem, willing to test alphas of improvements here. Currently I've hacked our test runner script to wait up to five seconds until file is available, but I sometimes even have crashes with that.

Agree with @andig here too that it would make sense for the `docker run` command to ensure consistent state (await mutagen sync?) before exiting fully. That would make any awaits in host scripts unnecessary since they would await the completion of `docker run` anyway.

**Comment by allenpetersen** [comment]
@andig agreed, that would be ideal. The proposal for `cache sync` was a MVP workaround.

I don't know how hard it would be for `docker run` to keep track of files it modified and wait for just those to sync.

What about this scenario:
Container A is long running and keep appending to a log file on a cached volume
Container B is short lived and creates a 50MB file.

Container B ideally wouldn't need to wait for container A's log file sync since it didn't modify that file.

**Comment by andig** [comment]
> What about this scenario:

Imho containers a and b are created by different run commands. Caching would need some kind of context (the run or other commands) and sync any changes done in that context only.

**Comment by djs55** [comment]
@allenpetersen @fdanielsen @andig : regarding the problem of synchronising files before/after running containers, I've made another prototype build with explicit `mutagen flush` on `docker run`. The experimental build is: https://desktop-stage.docker.com/mac/edge/45977/Docker.dmg I've also created an index of these builds in a public gist as it's getting a little bit difficult to keep track of them: https://gist.github.com/djs55/0e8957e1e36431a43cbfd3caaffa9165

Using this new build I can do this reliably:
```
# create a file in a container and then use it immediately from the host
$ docker run -it -v ~/workspace:/foo -w /foo alpine touch new-file && ls ~/workspace/new-file
/Users/djs/workspace/new-file

# create a file on the host and then immediately use it in a container
$ touch ~/workspace/host-file && docker run -it -v ~/workspace:/foo -w /foo alpine stat host-file
  File: host-file
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: fe01h/65025d	Inode: 2752523     Links: 1
Access: (0666/-rw-rw-rw-)  Uid: (  501/ UNKNOWN)   Gid: (   20/ dialout)
Access: 2020-06-16 12:31:11.000000000
Modify: 2020-06-16 12:31:11.000000000
Change: 2020-06-16 12:31:11.000000000
```
Let me know what you think!

**Comment by djs55** [comment]
@allenpetersen @fdanielsen @andig I've made another prototype build which explicitly flushes file changes also around `docker exec`: https://desktop-stage.docker.com/mac/edge/46638/Docker.dmg . This build also respects `.dockerignore` (for avoiding syncing certain directories) and allows sync to be bypassed by using the flag `:cached` (which will behave the same as `:delegated` did before). Let me know what you think!

**Comment by fdanielsen** [comment]
@djs55 I've been on leave after becoming a first time dad and a brief vacation, so haven't been able to test this build yet. But I see the edge channel for Docker desktop has had a couple updates in late July that includes some sync/mutagen improvements. Are these flush fixes also in the edge channel now, or is it still relevant to test these special builds? It seems the build number of the last build you posted is 46638, whilst the most recent Docker edge build I have installed is build 46980, so it seems more recent?
